### DISCUSSION TITLE
Skip specific mutation value in s_bytes primitive

### DISCUSSION THREAD

**Discussion by milanraf** [jtpereyda/boofuzz#665] [Ideas]
Hello to the Community

I am trying to figure out how to add a custom option in s_bytes primitive.
Since I am fuzzing against a device by reading and writing registers in its memory, I would like to skip a specific register address which would change the device IP and make it unreachable, blocking the fuzzing process.
One s_block of the payload has one entry like this:
`s_bytes(value=bytes([0x00, 0x00]), size=2, max_len=2, name="reg_addr")`
The idea is to add an extra parameter like:
`s_bytes(value=b"", size=None, padding=b"\x00", fuzzable=True, max_len=None, name=None, forbidden_val=b""):`
I would like to ask the Community where to look for in the code in order to try to implement this.
I have already found a candidate function:
`def get_mutations(self):`
in fuzzable.py
but it is not really clear how this recursive algorithm works.

Thanks for hints

Best

Raffaello

**Comment by cq674350529** [comment]
Hi, if you want to add extra parameters for `s_bytes()`, maybe you can refer to the implementation of `s_bytes()`. Based on what you described, you can try to add custom changes to the following methods.

The `mutations()` will call `self._iterate_fuzz_cases()` to get the mutated data.
https://github.com/jtpereyda/boofuzz/blob/63406302badae78a00619bcacf24c529f0c41957/boofuzz/primitives/bytes.py#L146-L151
https://github.com/jtpereyda/boofuzz/blob/63406302badae78a00619bcacf24c529f0c41957/boofuzz/primitives/bytes.py#L164-L179

**Comment by milanraf** [comment]
Thank you for the hint!
It's been a pretty trivial modification:
```
    def mutations(self, default_value):
        for fuzz_value in self._iterate_fuzz_cases(default_value):
            if fuzz_value != self.forbidden_value:
                if callable(fuzz_value):
                    yield compose(self._adjust_mutation_for_size, fuzz_value)
                else:
                    yield self._adjust_mutation_for_size(fuzz_value=fuzz_value)
```

**Reply by jtpereyda** [reply]
Love it! It could be made into a generic parameter, with some tweaking of base classes, but mostly I'm glad to hear it was easy enough to hack together. Happy fuzzing!
